### PR TITLE
Added log package

### LINKED ISSUE #76: Improve logger for User and Dev cases
In order to better manage log messages we need to separate the information the user needs from the detailed or verbose information that the developer needs. So that may require log level implementation. User necessary messages should go directly to standard out on the screen for the user to read. Same as debugging information if that level is set.

### PR DESCRIPTION
**Pull request by yuriolisa**
fixes #76 
Signed-off-by: Yuri Sa <yurimsa@gmail.com>
* [x] Added option `OPCAP_LOG` which can be set in one of these values `warn`, `info`, `error`, `debug`. 
* [x] Added flag `--log-level`. 
* [x] Added Zap pkg, implementing the Sugared logger.
* [x] Changed log messages to the respective log level prefix.

### CODE REVIEW

**Comment by yuriolisa** [comment]
Resolves #76

File: internal/logger/logger.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
+package logging
+
+import (
+	"fmt"
+	hclog "github.com/hashicorp/go-hclog"
```
Comment (acmenezes) [inline internal/logger/logger.go:5]: @yuriolisa I see you removed Zap, what are the advantages of hclog if compared with zap logger? We chose zap for a reason. It's very solid and widespread, has much bigger community and has bench mark tests for performance if we want to export logs in the future to another system. Besides that, that's the basis for Kubernetes logging in general and controller-runtime in specific. Could we accomplish the same with zap?

**Review by acmenezes — CHANGES_REQUESTED** [review]
Hi @yuriolisa, thanks for the PR! The work here looks good but can we stick to zap? I would change my vote here if I can see clear advantages on the hclog library and/or the majority of the team thinks we should go differently.

File: internal/logger/logger.go (hunk 1 above)
Comment (yuriolisa) [reply internal/logger/logger.go]: @acmenezes, I think the `hc-logger` it's simpler to implement, however, I agree with you when we advocate in favor to use the same Kubernetes basis logger and it's undoubtedly faster. Therefore, I switched the PR back to Zap package reading from an environment variable, called: `OPCAP_LOG`.

File: internal/logger/logger.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+type EncoderCfg struct {
+	MessageKey   string `json:"messageKey"`
+	LevelKey     string `json:"levelKey"`
+	LevelEncoder string `json:"levelEncoder"`
+}
+
+
+func InitLogger() *zap.Logger {
+	data := &LogJSON{
+		Level:       envLog,
```
Comment (yashoza19) [inline internal/logger/logger.go:31]: Instead on `envLog` can we implement this as a flag and allow the user to set envVar using the Flag itself? The Flag can be implemented at project level.

File: internal/operator/client.go [hunk 3]
```diff
@@ -82,7 +81,7 @@ func NewOpCapClient() (Client, error) {
 
 	client, err := runtimeClient.New(kubeConfig(), runtimeClient.Options{Scheme: scheme})
 	if err != nil {
-		logger.Errorf("could not get subscription client")
+		zap.S().Info("backend/local: apply calling Apply")
```
Comment (yashoza19) [inline internal/operator/client.go:84]: Can we change this to `logger`or `log` instead of `zap.S()`? Same comment for all changes.

Comment (yuriolisa) [reply internal/operator/client.go:84]: I've followed this way to avoid redeclaring the zap Logger for each function. I've changed the main.go, calling the Zap as Global Sugared Logger.

File: internal/logger/logger.go (hunk 2 above)
Comment (yashoza19) [reply internal/logger/logger.go:31]: Exactly it is being populated by an environment variable and not a flag. The idea is to implement a flag at `root.go` and it can also accept an envVar from there. Example:
```
flags.StringVar(&uploadflags.Trace, "trace", envy.Get("TRACE", "false"), "enable tracing; false by default")
```

File: cmd/check.go [hunk 4]
```diff
@@ -88,7 +90,7 @@ func init() {
 	flags.StringVar(&checkflags.CatalogSourceNamespace, "catalogsourcenamespace", "openshift-marketplace",
 		"specifies the namespace where the catalogsource exists")
 	flags.StringSliceVar(&checkflags.AuditPlan, "auditplan", defaultAuditPlan, "audit plan is the ordered list of operator test functions to be called during a capability audit.")
-	flags.BoolVar(&checkflags.ListPackages, "list-packages", false, "list packages in the catalog")
-	flags.StringSliceVar(&checkflags.FilterPackages, "filter-packages", []string{}, "a list of package(s) which limits audits and/or other flag(s) output")
-	flags.BoolVar(&checkflags.AllInstallModes, "all-installmodes", false, "when set, all install modes supported by an operator will be tested")
+	flags.BoolVar(&checkflags.ListPackages, "list-packages", false, "list packages in the catalog.")
```
Comment (bcrochet) [inline cmd/check.go:93]: Should these have periods at the end? I wouldn't think so...

File: internal/logger/logger.go (hunk 2 above)
Comment (yashoza19) [reply internal/logger/logger.go:31]: This way It can either be set through a flag or set through an env Var.

File: internal/capability/auditor.go [hunk 5]
```diff
@@ -21,31 +22,32 @@ type capAuditor struct {
...
 	return auditor, nil
 }
 
 // BuildWorkQueueByCatalog fills in the auditor workqueue with all package information found in a specific catalog
 func (capAuditor *capAuditor) BuildWorkQueueByCatalog(options operator.OperatorCheckOptions) error {
-
 	c, err := operator.NewOpCapClient()
 	if err != nil {
 		// if it doesn't load the client nothing can be done
 		// log and panic
```
Comment (bcrochet) [inline internal/capability/auditor.go:43]: Inaccurate comment now... (For the better I might add)

File: internal/capability/operand_cleanup.go [hunk 6]
```diff
@@ -2,20 +2,19 @@ package capability
...
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime/schema"
-)
+	"go.uber.org/zap"
+)	
 
 // OperandCleanup removes the operand from the OCP cluster in the ca.namespace
 func (ca *CapAudit) OperandCleanUp() error {
-	logger.Debugw("cleaningUp operand for operator", "package", ca.Subscription.Package, "channel", ca.Subscription.Channel, "installmode", ca.Subscription.InstallModeType)
+	zap.S().Debug("CleaningUp operand for operator", "package", ca.Subscription.Package, "channel", ca.Subscription.Channel, "installmode", ca.Subscription.InstallModeType)
```
Comment (bcrochet) [inline internal/capability/operand_cleanup.go:17]: This was cleaned up (ha!) in @acmenezes #188 patch. Should be "cleaning up"

File: internal/capability/operator_install.go [hunk 7]
```diff
@@ -2,16 +2,14 @@ package capability
...
 	"github.com/opdev/opcap/internal/operator"
+	"go.uber.org/zap"
 )
 
-var logger = log.Sugar
 
 func (ca *CapAudit) OperatorInstall() error {
-	logger.Debugw("installing package", "package", ca.Subscription.Package, "channel", ca.Subscription.Channel, "installmode", ca.Subscription.InstallModeType)
-
+
```
Comment (bcrochet) [inline internal/capability/operator_install.go:11]: Unwanted newline

File: main.go [hunk 8]
```diff
@@ -1,7 +1,16 @@
...
 
-import "github.com/opdev/opcap/cmd"
-
+import (
+	"github.com/opdev/opcap/cmd"
+	"go.uber.org/zap"
+	log "github.com/opdev/opcap/internal/logger"
+)
 func main() {
+	mainLogger := log.InitLogger()
```
Comment (bcrochet) [inline main.go:9]: Highly against any of this.... This does not belong in main(). Should be in a PersistentPreRunE in root.go.

File: internal/logger/logger.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+import (
+	"encoding/json"
+	"go.uber.org/zap"
+	"log"
+	"os"
+)
+
+var (
+	envLog  = os.Getenv("OPCAP_LOG")
```
Comment (bcrochet) [inline internal/logger/logger.go:11]:
```suggestion
	envLog  = os.Getenv("OPCAP_LOG_LEVEL")
```

**Review by bcrochet — CHANGES_REQUESTED** [review]
Comments and suggestions.

File: internal/operator/client.go (hunk 3 above)
Comment (yuriolisa) [reply internal/operator/client.go]: I've just applied the looger.

File: cmd/check.go (hunk 4 above)
Comment (yuriolisa) [reply cmd/check.go]: I've removed them.

File: internal/capability/operand_cleanup.go (hunk 6 above)
Comment (yuriolisa) [reply internal/capability/operand_cleanup.go]: Thanks, after the rebase got fixed.

File: main.go (hunk 8 above)
Comment (yuriolisa) [reply main.go]: Removed and initialized all the stuff inside the logger package.

File: .golangci.yaml [hunk 10]
```diff
@@ -0,0 +1,23 @@
+run:
```
Comment (bcrochet) [inline .golangci.yaml:1]: This file does not appear to be a part of this patch.

File: cmd/check.go [hunk 11]
```diff
@@ -8,18 +8,19 @@ import (
...
 	"github.com/opdev/opcap/internal/operator"
+	"github.com/opdev/opcap/internal/logger"
 
 	pkgserverv1 "github.com/operator-framework/operator-lifecycle-manager/pkg/package-server/apis/operators/v1"
 
 	"github.com/spf13/cobra"
 )
 
-// TODO: provide godoc compatible comment for checkCmd
+// TODO: provide godoc compatible comment for checkCmd.
```
Comment (bcrochet) [inline cmd/check.go:18]: For one, these should probably just be removed. But they don't appear to be related to this patch. And whatever it is that added a period you should probably turn off. :)

File: cmd/check.go [hunk 12]
```diff
@@ -70,8 +71,10 @@ Flags:
 			Packages:               checkflags.Packages,
 		}
 
-		// run all dynamically built audits in the auditor workqueue
-		capAuditor.RunAudits()
+		// run all dynamically built audits in the auditor workqueue.
+		if err := capAuditor.RunAudits(); err != nil {
```
Comment (bcrochet) [inline cmd/check.go:75]: Rather than add a block just to add a debug log (especially a debug log that says "error"), it should probably just be a RunE and do a `return capAuditor.RunAudits()`.

File: cmd/root.go [hunk 13]
```diff
@@ -7,7 +7,7 @@ import (
 	"github.com/spf13/cobra"
 )
 
-// rootCmd represents the base command when called without any subcommands
+// rootCmd represents the base command when called without any subcommands.
```
Comment (bcrochet) [inline cmd/root.go:10]: The periods again.

**Review by bcrochet — CHANGES_REQUESTED** [review]
I got just a few files in, and there are way to many extraneous changes (periods added to the end of comments seems to be the biggest culprit) to continue a review. Will look again once all of those are removed. And please keep this to a single commit in the PR. (Squash then amend from then on)

File: internal/logger/logger.go [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+	sugarLogger = logger.Sugar()
+}
+
+func Info(message string, fields ...interface{}) {
+	sugarLogger.Info(message, fields)
+}
+func Debug(message string, fields ...interface{}) {
+	sugarLogger.Debug(message, fields)
+}
+func Error(message string, fields ...interface{}) {
```
Comment (bcrochet) [inline internal/logger/logger.go:64]: Opinion: These should remain as Errorf, Debugf, etc. as that is a convention when a format string is used, and a variable number of args for the format.

Comment (yuriolisa) [reply internal/logger/logger.go:70]: I agree with your suggestion, doing it. Thank you.

File: cmd/check.go (hunk 12 above)
Comment (bcrochet) [reply cmd/check.go:75]: Not a blocker.

File: internal/capability/operand_install.go [hunk 15]
```diff
@@ -53,7 +53,10 @@ func (ca *capAudit) getAlmExamples() error {
 func (ca *capAudit) OperandInstall() error {
 	logger.Debugw("installing operand for operator", "package", ca.subscription.Package, "channel", ca.subscription.Channel, "installmode", ca.subscription.InstallModeType)
 
-	ca.getAlmExamples()
+	err := ca.getAlmExamples()
```
Comment (bcrochet) [inline internal/capability/operand_install.go:56]: Non-blocking: These could be one line.

File: internal/capability/operator_install.go [hunk 16]
```diff
@@ -21,12 +19,18 @@ func (ca *capAudit) OperatorInstall() error {
 	// create remaining target namespaces watched by the operator
 	for _, ns := range ca.operatorGroupData.TargetNamespaces {
 		if ns != ca.namespace {
-			operator.CreateNamespace(context.Background(), ns)
+			_, err := operator.CreateNamespace(context.Background(), ns)
```
Comment (bcrochet) [inline internal/capability/operator_install.go:22]: Non-blocking: If you're ignoring the first param, these could be a single line.

**Review by bcrochet — COMMENTED** [review]
Just a couple of non-blocker comments.

This really should just have a single commit as well.

File: cmd/check.go [hunk 17]
```diff
@@ -99,4 +105,5 @@ func init() {
 	flags.StringSliceVar(&checkflags.AuditPlan, "audit-plan", defaultAuditPlan, "audit plan is the ordered list of operator test functions to be called during a capability audit.")
 	flags.BoolVar(&checkflags.ListPackages, "list-packages", false, "list packages in the catalog")
 	flags.StringSliceVar(&checkflags.Packages, "packages", []string{}, "a list of package(s) which limits audits and/or other flag(s) output")
+	flags.StringVar(&checkflags.LogLevel, "log-level", "", "specifies the one of the log levels in order of decreasing verbosity: debug, error, info, warn")
```
Comment (yashoza19) [inline cmd/check.go:108]: I was suggesting this to be a root-level flag and not a check subcommand flag. But that might complicate things a bit. @bcrochet @acmenezes any thoughts?

Comment (acmenezes) [reply cmd/check.go:108]: @yashoza19 I guess it may need to be root level and persistent to be available under any command. But I don't think it's a blocker since check is the command that most requires debugging for now. So, there is a lot of value adding it as it is. We can open an issue to optimize flags in the near future since other commands will land pretty soon.

**Review by acmenezes — APPROVED** [review]

File: internal/logger/logger.go [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+	if err != nil {
+		panic(err)
+	}
+	sugarLogger = logger.Sugar()
+}
+//Info exports Info Suggared Loglevel.
+func Info(message string, fields ...interface{}) {
+	sugarLogger.Info(message, fields)
+}
+//Debug exports Suggared Loglevel.
```
Comment (yashoza19) [inline internal/logger/logger.go:68]: A couple of small nits: 

- start comment with lower case
- leave space before starting a comment
- remove period for consistency across all comment

File: internal/logger/logger.go [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+//Debugw exports Suggared Loglevel.
+func Debugw(message string, fields ...interface{}) {
+	sugarLogger.Debugw(message, fields)
+}
+//Debugf exports Suggared Loglevel.
+func Debugf(message string, fields ...interface{}) {
+	sugarLogger.Debugf(message, fields)
+}
+//Error exports Suggared Loglevel.
+func Error(message string, fields ...interface{}) {
```
Comment (bcrochet) [inline internal/logger/logger.go:81]: Why are we keeping these? At least like this? 

If you want to keep them, that's fine, but they should just take a string, and call the local 'f' version with no additional args.

Comment (yuriolisa) [reply internal/logger/logger.go]: I have just removed it.

**Review by bcrochet — CHANGES_REQUESTED** [review]
Seems like the merge commits are still present. This really needs to be collapsed to a single commit. And it looks like a `make tidy` is necessary.

File: internal/capability/operand_install.go [hunk 20]
```diff
@@ -118,8 +120,15 @@ func (ca *capAudit) OperandInstall(ctx context.Context) error {
 	}
 	defer file.Close()
 
-	ca.OperandInstallJsonReport(file)
-	ca.OperandTextReport(os.Stdout)
+	if err := ca.OperandInstallJsonReport(file); err != nil {
+		logger.Debugf("Error during the OperatorInstall Report: %w", err)
```
Comment (yashoza19) [inline internal/capability/operand_install.go:124]: Fix typos: "Error during the OperatorInstall Report: %w" to "Error during the OperandInstall"

File: internal/capability/operand_install.go [hunk 21]
```diff
@@ -118,8 +120,15 @@ func (ca *capAudit) OperandInstall(ctx context.Context) error {
...
+		return err
+	}
+
+	if err := ca.OperandTextReport(os.Stdout); err != nil {
+		logger.Debugf("Error during the OperatorInstall Text Report: %w", err)
```
Comment (yashoza19) [inline internal/capability/operand_install.go:129]: Fix typos: "Error during the OperatorInstall Report: %w" to "Error during the OperandInstall"

File: cmd/check.go [hunk 22]
```diff
@@ -92,4 +95,6 @@ func init() {
 	flags.BoolVar(&checkflags.ListPackages, "list-packages", false, "list packages in the catalog")
 	flags.StringSliceVar(&checkflags.Packages, "packages", []string{}, "a list of package(s) which limits audits and/or other flag(s) output")
 	flags.BoolVar(&checkflags.AllInstallModes, "all-installmodes", false, "when set, all install modes supported by an operator will be tested")
+	flags.StringVar(&checkflags.LogLevel, "log-level", "", "specifies the one of the log levels in order of decreasing verbosity: debug, error, info, warn")
+	flags.BoolVar(&checkflags.AllInstallModes, "all-installmodes", false, "when set, all install modes supported by an operator will be tested")
```
Comment (yashoza19) [inline cmd/check.go:99]: This is repeating twice.

**Review by yashoza19 — CHANGES_REQUESTED** [review]

**Review by yashoza19 — APPROVED** [review]
/approve

**Review by acmenezes — APPROVED** [review]
Looks great now.

**Review by skattoju — APPROVED** [review]
/lgtm

**Review by skattoju — APPROVED** [review]
/approve
